test/scripts/parallel-local/main.js

```diff
@@ -0,0 +1,234 @@
+    }
+};
+
+/**
+ * Sets up event listeners for a worker process.
+ * @param {ChildProcess} worker - The worker process.
+ * @param {number} port - The port number assigned to the worker.
+ * @param {string} projectName - The project name assigned to the worker.
+ * @returns {void} Does not return a value.
+ */
+const setupWorkerListeners = (worker, port, projectName) => {
+    worker.stdout.on('data', (data) => console.log(`${projectName}: ${data.toString()}`));
+    worker.stderr.on('data', (data) => console.error(`${projectName} Error: ${data.toString()}`));
+
+    worker.on('message', (msg) => handleWorkerMessage(msg, worker, projectName));
+    worker.on('exit', (code) => handleWorkerExit(code, projectName));
+};
+
+/**
+ * Handles messages from worker processes, potentially assigning new tests.
+ * @param {string} msg - The message from the worker.
+ * @param {ChildProcess} worker - The worker process.
+ * @param {string} projectName - The project name.
+ * @returns {void} Does not return a value.
+ */
+const handleWorkerMessage = (msg, worker, projectName) => {
+    if (msg === 'request_next_test') {
+        assignTestToWorker(worker, projectName);
+    }
+};
+
+/**
+ * Assigns a test to a worker and logs the activity.
+ * @param {ChildProcess} worker - The worker process.
+ * @param {string} projectName - The project name.
+ * @returns {void} Does not return a value.
+ */
+const assignTestToWorker = (worker, projectName) => {
+    if (remainingTests.length > 0) {
+        const test = remainingTests.pop();
+        console.log(`${projectName} starting new suite: ${test} (${testSuites.length - remainingTests.length}/${testSuites.length})`);
+        worker.send({ test, projectName });
+    } else {
+        console.log(`${projectName} found no more tests...`);
+        worker.send('no_more_tests');
+        workersCompleted.add(projectName);
+    }
+};
+
+/**
+ * Logs the exit of worker processes and checks for completion.
+ * @param {number} code - The exit code of the worker process.
+ * @param {string} projectName - The project name.
+ * @returns {void} Does not return a value.
+ */
+const handleWorkerExit = (code, projectName) => {
+    if (code !== 0) {
+        console.error(`Worker ${projectName} exited with code ${code}`);
+    } else {
+        console.log(`Worker ${projectName} completed successfully`);
+    }
+    workersExited.add(projectName);
+    if (workersCompleted.size === numWorkers && workersExited.size === numWorkers) {
+        displayResults();
+    }
+};
+
+/**
+ * Displays the results of all tests by reading specific log files and aggregating results.
+ * It first prints the contents of 'tests.log', calculates totals from 'results.log',
+ * and finally reads 'fails.log' for any failures.
+ * @returns {void} Does not return a value.
+ */
+const displayResults = () => {
+    console.log('\nResults:\n');
+    testSuites.forEach((testSuiteName) => {
+        readLogFiles(testSuiteName, 'tests.log');
+    });
```

Comment:
You should try as much as possible to use async versions of files handling tools.
Here for example your function should return a promise, and you should use Promise.all on all of them to get the array of files, and them concatenate it if needed to have it in order.

Reply:
Done